File: ros_time_example/scripts/timer_example.py
#! /usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.time import Time, Duration


class TimerExample(Node):
    def __init__(self):
        super().__init__('timer_example')
        self.timer = self.create_timer(timer_period_sec=0.1, callback=self.timer_callback)
        self.last_time = Time()

    def timer_callback(self):
        if self.last_time.clock_type != self.get_clock().now().clock_type:
            self.last_time = self.get_clock().now()
            return

        current_time = self.get_clock().now()
        time_diff = Duration()
        time_diff = current_time - self.last_time
        self.last_time = current_time

        self.get_logger().info(f'Time since last trigger: {1e-9 * time_diff.nanoseconds}')


if __name__ == '__main__':
    rclpy.init()
    node_instance = TimerExample()
    rclpy.spin(node_instance)

File: ros_time_example/src/timer_example.cpp
#include <rclcpp/rclcpp.hpp>

namespace ros_time_example {

  class TimerExample : public rclcpp::Node {
    public:
      TimerExample()
      : rclcpp::Node("timer_example")
      {
        timer_ = rclcpp::create_timer(this, get_clock(), std::chrono::milliseconds(1000),
          std::bind(&TimerExample::timer_callback, this));
      }

    private:
      rclcpp::TimerBase::SharedPtr timer_;
      rclcpp::Time last_time_;

      void timer_callback() {
        if (last_time_.get_clock_type() != get_clock()->now().get_clock_type()) {
          last_time_ = get_clock()->now();
          return;
        }

        rclcpp::Time current_time = get_clock()->now();
        rclcpp::Duration time_diff = current_time - last_time_;
        last_time_ = current_time;

        RCLCPP_INFO_STREAM(get_logger(), "Time since last trigger: " << time_diff.seconds());
      }
  };
}

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  auto node_instance = std::make_shared<ros_time_example::TimerExample>();
  rclcpp::spin(node_instance);
  rclcpp::shutdown();
  return 0;
}
